packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts

```diff
@@ -0,0 +1,199 @@
+    const credential =
+      users.list['joe@example.com'].user.email +
+      ':' +
+      users.list['joe@example.com'].user.password;
+    const hash = Buffer.from(credential).toString('base64');
+    await client
+      .get('/whoAmI')
+      .set('Authorization', 'NotB@s1c ' + hash)
+      .expect(401);
+  });
+
+  it(`returns error for missing ':' in decrypted Authorization header credentials value`, async () => {
+    const client = whenIMakeRequestTo(server);
+    const credential =
+      users.list['joe@example.com'].user.email +
+      '|' +
+      users.list['joe@example.com'].user.password; // substituting ':' with '|'
+    const hash = Buffer.from(credential).toString('base64');
+    await client
+      .get('/whoAmI')
+      .set('Authorization', 'Basic ' + hash)
+      .expect(401);
+  });
+
+  it(`returns error for too many parts in decrypted Authorization header credentials value`, async () => {
+    const client = whenIMakeRequestTo(server);
+    const credential =
+      users.list['joe@example.com'].user.email +
+      ':' +
+      users.list['joe@example.com'].user.password +
+      ':' +
+      'extraPart'; // three parts instead of two
+    const hash = Buffer.from(credential).toString('base64');
+    await client
+      .get('/whoAmI')
+      .set('Authorization', 'Basic ' + hash)
+      .expect(401);
+  });
+
+  it('allows anonymous requests to methods with no decorator', async () => {
+    class InfoController {
+      @get('/status')
+      status() {
+        return {running: true};
+      }
+    }
+
+    app.controller(InfoController);
+    await whenIMakeRequestTo(server)
+      .get('/status')
+      .expect(200, {running: true});
+  });
+
+  async function givenAServer() {
+    app = getApp();
+    server = await app.getServer(RestServer);
+  }
+
+  function givenControllerInApp() {
+    const apispec = anOpenApiSpec()
+      .withOperation('get', '/whoAmI', {
+        'x-operation-name': 'whoAmI',
+        responses: {
+          '200': {
+            description: '',
+            schema: {
+              type: 'string',
+            },
+          },
+        },
+      })
+      .build();
+
+    @api(apispec)
+    class MyController {
+      constructor(
+        @inject(AuthenticationBindings.CURRENT_USER) private user: UserProfile,
+      ) {}
+
+      @authenticate('basic')
+      async whoAmI(): Promise<string> {
+        if (this.user) {
+          if (this.user.email) return this.user.email;
+          else return 'user email is undefined';
+        } //if
+        else return 'user is undefined';
+      }
+    }
+
+    app.controller(MyController);
+  }
+
+  it('returns error for unknown authentication strategy', async () => {
+    class InfoController {
+      @get('/status')
+      @authenticate('doesnotexist')
+      status() {
+        return {running: true};
+      }
+    }
+
+    app.controller(InfoController);
+    await whenIMakeRequestTo(server)
+      .get('/status')
+      .expect(401);
```

Comment:
Should we use a different http status code when a strategy is not available?

Reply:
regarding
```ts
it('returns error for unknown authentication strategy', async () => {
    class InfoController {
      @get('/status')
      @authenticate('doesnotexist')
      status() {
        return {running: true};
      }
    }

    app.controller(InfoController);
    await whenIMakeRequestTo(server)
      .get('/status')
      .expect(401);
  });
```
If the user decorates the controller method with the `@authenticate` decorator, then this means he/she wants the endpoint protected. If the user specifies a `name` that he/she hasn't registered as an extension, then I strongly believe the user should receive a `401 - Not Authorized`. If we don't throw this error, this will allow access the endpoint they were trying to protect...this isn't good.

If something goes wrong through any code that is helping with authenticating an endpoint, I am returning a 401 Not authorized in my mock strategy implementations and mock user service and token service.

Reply:
Plus... we don't ship a custom sequence with the authentication action...the user has to create one. They can decide in the sequence code block , what error code to return. It just so happens I decided to return 401...but they can decide otherwise.